ts/private/ts_project_worker.js

```diff
@@ -403,13 +404,13 @@ function createProgram(args, initialInputs) {
     }
 
     function clearTimeout(i) {
-        taskQueue[i] = null;
+        delete taskQueue[i];
     }
 
     function applyChanges() {
-        debuglog(`Applying changes ${taskQueue.length}`);
-        while(taskQueue.length) {
-            const task = taskQueue.shift();
+        for (let i = 0; i < taskQueue.length; i++) {
+            const task = taskQueue[i];
+            delete taskQueue[i]
```

Comment:
Why delete instead of `pop` or `shift` like before? This means the array length won't change and it will grow forever?

Reply:
if we pop or shift we'd be changing the index of the task array. so when a new task is scheduled (setTimeout) we push the task into the array and return back the index to be used later by tsc to clean it up (clearTimeout). we throw out the task array after each apply

Reply:
I see now that I look around more... I do wonder if we should do this differently though...

Reply:
I'd be interested to see different fake scheduling algorithms. we don't need anything fancy though. there are only two things that tsc schedules and both of them have the same `timeout` value hence no need to worry about task urgency.

Reply:
First thing that comes to mind would be:
```
let nextTimerId = 0;
function setTimeout(cb) { taskQueue[++nextTimerId] = cb; return nextTimerId; }
function clearTimeout(id) { delete taskQueue[id] }
function applyChanges() {
   for (const id of taskQueue) {
        taskQueue[id]?();
        delete taskQueue[id]
    }
}
```

However that looses the order which I assume is important, so... 🤔 

```
let nextTimerId = 1;
let lastRun = 0;
function setTimeout(cb) { taskQueue[++nextTimerId] = cb; return nextTimerId; }
function clearTimeout(id) { delete taskQueue[id] }
function applyChanges() {
   while (lastRun++ < nextTimerId)
        taskQueue[lastRun]?();
        delete taskQueue[lastRun]
    }
}
```

Reply:
the second one could work. do you believe this is better in terms of readability? I didn't like the use of lastRun. kind of makes it harder to understand.

Reply:
You're essentially doing the exact same thing today except all the data is hidden in the array. The `nextTimerId` is the `taskQueue.length` and the `lastRun` is `taskQueue.findIndex(e => !!e)`.

Note that today the array grows forever so when you do the loop the first `lastRun-1` entries will be undefined. The array also grows forever.

Reply:
maybe we should just put `taskQueue = new Array(1)` to the last line of applyChanges so it doesn't grow indefinitely?

Reply:
Doesn't that just continue making it more complicated? Then the `new Array(1)` thing is in multiple places. I was hoping to remove it...

Reply:
We can think about all this stuff later though, don't have to do it in this PR...

Reply:
added logic to clear the array after each apply.